staging/src/k8s.io/apiserver/pkg/storage/storagecluster/tenantstoragemapmanager.go

```diff
@@ -0,0 +1,273 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storagecluster
+
+import (
+	"errors"
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/diff"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+	"strconv"
+	"sync"
+)
+
+var instanceTenantMapManager *TenantStorageMapManager
+
+type tenantStorage struct {
+	clusterId string
+	rev       int64
+}
+
+type TenantStorageMapManager struct {
+	tenantListerSynced cache.InformerSynced
+	tenantLister       corelisters.TenantLister
+
+	tenantToClusterMap map[string]*tenantStorage
+
+	mux sync.RWMutex
+	rev int64
+}
+
+var GetClusterIdFromTenantHandler = getClusterIdFromTenant
+
+func GetTenantStorageManager() *TenantStorageMapManager {
+	return instanceTenantMapManager
+}
+
+func checkTenantStorageInstanceExistence() {
+	if instanceTenantMapManager != nil {
+		klog.Fatalf("Unexpected reference to tenant storage cluster manager - initialized")
+	}
+}
+
+func NewTenantStorageMapManager(tenantInformer coreinformers.TenantInformer) *TenantStorageMapManager {
+	checkTenantStorageInstanceExistence()
+
+	manager := &TenantStorageMapManager{
+		tenantToClusterMap: make(map[string]*tenantStorage),
+		rev:                int64(0),
+	}
+
+	tenantInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    manager.addTenant,
+		UpdateFunc: manager.updateTenant,
+		DeleteFunc: manager.deleteTenant,
+	})
+
+	manager.tenantLister = tenantInformer.Lister()
+	manager.tenantListerSynced = tenantInformer.Informer().HasSynced
+	err := manager.syncTenants()
+	if err != nil {
+		klog.Fatalf("Unable to get tenant storage from registry. Error %v", err)
+	}
+
+	instanceTenantMapManager = manager
+	return instanceTenantMapManager
+}
+
+func (ts *TenantStorageMapManager) Run(stopCh <-chan struct{}) {
+	klog.Infof("Starting tenant storage map manager.")
+	defer klog.Infof("Shutting down tenant storage map manager.")
+
+	if !cache.WaitForCacheSync(stopCh, ts.tenantListerSynced) {
+		klog.Infof("Tenant storage map NOT synced %v.", ts.tenantListerSynced)
+		return
+	}
+
+	klog.Infof("Caches are synced for tenant storage map.")
+	<-stopCh
+}
+
+func (ts *TenantStorageMapManager) syncTenants() error {
+	ts.mux.Lock()
+	klog.V(4).Infof("mux acquired syncTenants.")
+	defer func() {
+		ts.mux.Unlock()
+		klog.V(4).Infof("mux released syncTenants.")
+	}()
+
+	// Question: do we need to ignore deleted tenants?
```

Comment:
Can we check there? Like deleteTimeStamp?

Reply:
I think list should not have objects that have deleteTimeStamp